src/ROOTDataSource.cc

```diff
@@ -0,0 +1,308 @@
+#include <iostream>
+#include <memory>
+
+// ROOT
+#include <TFile.h>
+
+// podio
+#include <podio/Frame.h>
+#include <podio/ROOTReader.h>
+
+namespace podio {
+ROOTDataSource::ROOTDataSource(const std::string& filePath, int nEvents) : m_nSlots{1} {
+  m_filePathList.emplace_back(filePath);
+  SetupInput(nEvents);
+}
+
+ROOTDataSource::ROOTDataSource(const std::vector<std::string>& filePathList, int nEvents) :
+    m_nSlots{1}, m_filePathList{filePathList} {
+  SetupInput(nEvents);
+}
+
+/// @TODO Check for the existence of the file, which might be coming from web
+///       or EOS.
+void ROOTDataSource::SetupInput(int nEvents) {
+  // std::cout << "podio::ROOTDataSource: Constructing the source ..." << std::endl;
+
+  if (m_filePathList.empty()) {
+    throw std::runtime_error("podio::ROOTDataSource: No input files provided!");
+  }
+
+  for (const auto& filePath : m_filePathList) {
+    // Check if file exists
+    // if (!std::filesystem::exists(filePath)) {
+    //   throw std::runtime_error("podio::ROOTDataSource: Provided file \""
+    //                            + filePath + "\" does not exist!");
+    // }
+
+    // Check if the provided file contains required metadata
+    std::unique_ptr<TFile> inFile(TFile::Open(filePath.data(), "READ"));
+    auto metadata = inFile->Get("podio_metadata");
+    if (!metadata) {
+      throw std::runtime_error("podio::ROOTDataSource: Provided file is missing podio metadata!");
+    }
+  }
+
+  // Create probing frame
+  podio::Frame frame;
+  unsigned int nEventsInFiles = 0;
+  podio::ROOTReader podioReader;
+  podioReader.openFiles(m_filePathList);
+  nEventsInFiles = podioReader.getEntries("events");
+  frame = podio::Frame(podioReader.readEntry("events", 0));
+
+  // Determine over how many events to run
+  if (nEventsInFiles > 0) {
+    /*
+    std::cout << "podio::ROOTDataSource: Found " << nEventsInFiles
+              << " events in files: \n";
+    for (const auto& filePath : m_filePathList) {
+      std::cout << "               - " << filePath << "\n";
+    }
+    */
+  } else {
+    throw std::runtime_error("podio::ROOTDataSource: No events found!");
+  }
+
+  if (nEvents < 0) {
+    m_nEvents = nEventsInFiles;
+  } else if (nEvents == 0) {
+    throw std::runtime_error("podio::ROOTDataSource: Requested to run over zero events!");
+  } else {
+    m_nEvents = nEvents;
+  }
+  if (nEventsInFiles < m_nEvents) {
+    m_nEvents = nEventsInFiles;
+  }
+
+  // std::cout << "podio::ROOTDataSource: Running over " << m_nEvents << " events."
+  //           << std::endl;
+
+  // Get collections stored in the files
+  std::vector<std::string> collNames = frame.getAvailableCollections();
+  // std::cout << "podio::ROOTDataSource: Found following collections:\n";
+  for (auto& collName : collNames) {
+    const podio::CollectionBase* coll = frame.get(collName);
+    if (coll->isValid()) {
+      m_columnNames.emplace_back(collName);
+      m_columnTypes.emplace_back(coll->getValueTypeName());
```

Comment:
```suggestion
      m_columnTypes.emplace_back(coll->getTypeName());
```
Since the only usage below is with `+ "Collection"`, we could just store the complete type name, including the collection here.

Reply:
Done